### DISCUSSION TITLE
is it possible to watch tracked env vars?

### DISCUSSION THREAD

**Discussion by MushiTheMoshi** [spf13/viper#1947] [Q&A]
Hi. greatly appreciated the awesome work done here.

I am wondering whether is possible to watch for changes over environment variables?

Regards,
Julio

**Comment by vynazevedo** [comment]
Viper doesn't have an environment-variable change callback. `WatchConfig` watches the config file through fsnotify.

It does reread the **current process's environment** on each getter call, though. Checked with Viper 1.21.0:

```go
v := viper.New()
if err := v.BindEnv("mode", "APP_MODE"); err != nil {
    panic(err)
}
if err := os.Setenv("APP_MODE", "before"); err != nil {
    panic(err)
}
fmt.Println(v.GetString("mode")) // before
if err := os.Setenv("APP_MODE", "after"); err != nil {
    panic(err)
}
fmt.Println(v.GetString("mode")) // after
```

Changing an `export` in a shell after starting the application won't change the running application's environment. Neither will editing a `.env` file unless something explicitly reloads it. Also, a value you've already copied into a variable or unmarshaled into a struct doesn't update itself.

If the goal is live configuration from outside the process, a config file plus `WatchConfig`/`OnConfigChange` fits that use case. For changes made with `os.Setenv` inside the process, you can read the getter again, but Viper won't notify you automatically.

Reference: [working with environment variables](https://github.com/spf13/viper#working-with-environment-variables).
